lib/codecs/src/encoding/format/protobuf.rs

```diff
@@ -0,0 +1,466 @@
+use std::path::PathBuf;
+use tokio_util::codec::Encoder;
+use vector_core::{
+    config::DataType,
+    event::{Event, Value},
+    schema,
+};
+
+/// Config used to build a `ProtobufSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone)]
+pub struct ProtobufSerializerConfig {
+    /// Options for the Protobuf serializer.
+    pub protobuf: ProtobufSerializerOptions,
+}
+
+impl ProtobufSerializerConfig {
+    /// Build the `ProtobufSerializer` from this configuration.
+    pub fn build(&self) -> Result<ProtobufSerializer, BuildError> {
+        let message_descriptor =
+            get_message_descriptor(&self.protobuf.desc_file, &self.protobuf.message_type)?;
+        Ok(ProtobufSerializer { message_descriptor })
+    }
+
+    /// The data type of events that are accepted by `ProtobufSerializer`.
+    pub fn input_type(&self) -> DataType {
+        DataType::Log.and(DataType::Trace)
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement(&self) -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // Protobuf, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Protobuf serializer options.
+#[crate::configurable_component]
+#[derive(Debug, Clone)]
+pub struct ProtobufSerializerOptions {
+    /// The path to the protobuf descriptor set file.
+    ///
+    /// This file is the output of `protoc -o <path> ...`
+    #[configurable(metadata(docs::examples = "/etc/vector/protobuf_descriptor_set.desc"))]
+    pub desc_file: PathBuf,
+
+    /// The name of the message type to use for serializing.
+    #[configurable(metadata(docs::examples = "package.Message"))]
+    pub message_type: String,
+}
+
+/// Serializer that converts an `Event` to bytes using the Protobuf format.
+#[derive(Debug, Clone)]
+pub struct ProtobufSerializer {
+    /// The protobuf message definition to use for serialization.
+    message_descriptor: MessageDescriptor,
+}
+
+/// Convert a single raw vector `Value` into a protobuf `Value`.
+///
+/// Unlike `convert_value`, this ignores any field metadata such as cardinality.
+fn convert_value_raw(
+    value: Value,
+    kind: &prost_reflect::Kind,
+) -> Result<prost_reflect::Value, vector_common::Error> {
+    let kind_str = value.kind_str().to_owned();
+    match (value, kind) {
+        (Value::Boolean(b), Kind::Bool) => Ok(prost_reflect::Value::Bool(b)),
+        (Value::Bytes(b), Kind::Bytes) => Ok(prost_reflect::Value::Bytes(b)),
+        (Value::Bytes(b), Kind::String) => Ok(prost_reflect::Value::String(
+            String::from_utf8_lossy(&b).into_owned(),
+        )),
+        (Value::Bytes(b), Kind::Enum(descriptor)) => {
+            let string = String::from_utf8_lossy(&b).into_owned();
+            // check for an exact enum name match
+            if let Some(d) = descriptor.values().filter(|v| v.name() == &string).next() {
```

Comment:
🥜 Can we merge this and the following filter?

Reply:
Oops, yep, I was under the impression that enum values were case sensitive, but they are not.  Merged.